File: physicar_bringup/scripts/topic_watchdog_node.py
#!/usr/bin/env python3
"""
Topic Watchdog Node for PhysiCar.

Monitors critical sensor topics. If a topic stops publishing for longer than
its configured timeout, kills the responsible process (launch will respawn it
because every node has respawn=True).

This handles cases where a node is alive (process still running, ROS spin
working) but its publish callback got stuck — common with libcamera, V4L2,
DDS shared-memory glitches, etc.

Restart policy:
- Startup grace period: ignore stale topics for the first N seconds after
  boot or after any kill (gives respawned nodes time to initialise).
- Cooldown: don't kill the same target twice within `cooldown` seconds.
"""

import os
import signal
import subprocess
import time

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy

from rosgraph_msgs.msg import Clock
from sensor_msgs.msg import CameraInfo, LaserScan, BatteryState, Imu
from nav_msgs.msg import Odometry


# topic, msg_type, timeout(s), kill_pattern (pgrep -f)
REAL_WATCH_TOPICS = [
    ("/camera/camera_info", CameraInfo, 5.0, "camera_ros/lib/camera_ros/camera_node"),
    ("/scan",                        LaserScan,        3.0, "rplidar_node"),
    ("/scan_filtered",               LaserScan,        3.0, "scan_filter_node"),
    ("/odom",                        Odometry,         3.0, "ekf_filter_node"),
    ("/battery_state",               BatteryState,    15.0, "physicar_driver_node"),
    ("/imu",                         Imu,              5.0, "physicar_driver_node"),
]

# SIM mode: watch rf2o, EKF, and Gazebo bridge topics.
# rf2o is the primary target: its Rate::sleep() blocks on sim time backward
# jumps after a world switch, stalling /odom/laser and cascading to EKF.
# /scan comes from ros_gz_bridge — if gz sim restarts, the bridge may lose
# its gz-transport connection and stop forwarding. Killing it triggers
# respawn (respawn=True in sim.launch.py), reconnecting to the new gz sim.
SIM_WATCH_TOPICS = [
    ("/scan",       LaserScan, 10.0, "ros_gz_bridge"),
    ("/odom/laser", Odometry,   5.0, "rf2o_laser_odometry_node"),
    ("/odom",       Odometry,  15.0, "ekf_filter_node"),
]


class TopicWatchdog(Node):
    def __init__(self):
        super().__init__("topic_watchdog")

        self.declare_parameter("startup_grace_sec", 30.0)
        self.declare_parameter("cooldown_sec", 30.0)
        self.declare_parameter("check_period_sec", 2.0)
        self.declare_parameter("enabled", True)
        self.declare_parameter("mode", "real")  # "real" or "sim"

        self.enabled = bool(self.get_parameter("enabled").value)
        if not self.enabled:
            self.get_logger().info("[Watchdog] disabled (enabled=False)")
            return

        mode = str(self.get_parameter("mode").value)
        self._topics = SIM_WATCH_TOPICS if mode == "sim" else REAL_WATCH_TOPICS

        self.startup_grace = float(self.get_parameter("startup_grace_sec").value)
        self.cooldown = float(self.get_parameter("cooldown_sec").value)
        self.check_period = float(self.get_parameter("check_period_sec").value)

        self.start_time = time.monotonic()
        self.last_msg_time: dict[str, float] = {}
        self.last_kill_time: dict[str, float] = {}
        self.subs = []

        qos_be = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            history=HistoryPolicy.KEEP_LAST,
            depth=1,
        )
        qos_rel = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST,
            depth=1,
        )

        for topic, msg_type, _timeout, _pattern in self._topics:
            qos = qos_be if msg_type in (LaserScan, Imu) else qos_rel
            self.last_msg_time[topic] = self.start_time
            self.subs.append(self.create_subscription(
                msg_type, topic,
                lambda _msg, t=topic: self._on_msg(t),
                qos,
            ))

        self.timer = self.create_timer(self.check_period, self._check)
        self.get_logger().info(
            f"[Watchdog] mode={mode}, monitoring {len(self._topics)} topics "
            f"(grace={self.startup_grace:.0f}s, cooldown={self.cooldown:.0f}s)"
        )

        # SIM mode: monitor /clock for backward time jumps (world switch).
        # When sim time jumps backward by >1s, immediately kill all watched
        # processes instead of waiting for stale timeouts.
        self._last_sim_sec = 0.0
        if mode == "sim":
            self.create_subscription(
                Clock, "/clock",
                self._on_clock,
                QoSProfile(
                    reliability=ReliabilityPolicy.BEST_EFFORT,
                    history=HistoryPolicy.KEEP_LAST, depth=1,
                ),
            )

    # ── Clock jump detection (sim only) ──

    def _on_clock(self, msg: Clock):
        sec = msg.clock.sec + msg.clock.nanosec * 1e-9
        prev = self._last_sim_sec
        self._last_sim_sec = sec

        # Ignore initial messages (prev == 0) and normal forward ticks
        if prev < 1.0 or sec >= prev:
            return

        # Backward jump detected (e.g. 120s → 0.5s) → world switch
        now = time.monotonic()

        # Respect startup grace
        if now - self.start_time < self.startup_grace:
            return

        self.get_logger().warn(
            f"[Watchdog] /clock jumped backward ({prev:.1f}s → {sec:.1f}s) "
            "— world switch detected, killing watched processes"
        )

        killed = set()
        for _topic, _msg_type, _timeout, pattern in self._topics:
            if pattern in killed:
                continue
            last_kill = self.last_kill_time.get(pattern, 0.0)
            if now - last_kill < self.cooldown:
                continue
            if self._kill(pattern):
                self.last_kill_time[pattern] = now
                killed.add(pattern)
        # Reset all topic timestamps so stale check doesn't re-trigger
        for t, _m, _to, _p in self._topics:
            self.last_msg_time[t] = now

    def _on_msg(self, topic: str):
        self.last_msg_time[topic] = time.monotonic()

    def _check(self):
        now = time.monotonic()

        if now - self.start_time < self.startup_grace:
            return

        for topic, _msg_type, timeout, pattern in self._topics:
            last = self.last_msg_time.get(topic, self.start_time)
            stale = now - last
            if stale < timeout:
                continue

            last_kill = self.last_kill_time.get(pattern, 0.0)
            if now - last_kill < self.cooldown:
                continue

            self.get_logger().warn(
                f"[Watchdog] {topic} stale for {stale:.1f}s "
                f"(timeout={timeout:.0f}s) → killing '{pattern}'"
            )
            if self._kill(pattern):
                self.last_kill_time[pattern] = now
                # reset all topics that share this pattern so respawn has time
                for t, _m, _to, p in self._topics:
                    if p == pattern:
                        self.last_msg_time[t] = now

    def _kill(self, pattern: str) -> bool:
        try:
            res = subprocess.run(
                ["pgrep", "-f", pattern],
                capture_output=True, text=True, timeout=3,
            )
            pids = [int(p) for p in res.stdout.split() if p.isdigit()]
            # Filter out our own process and the launch parent (defensive)
            my_pid = os.getpid()
            pids = [p for p in pids if p != my_pid]
            if not pids:
                self.get_logger().warn(f"[Watchdog] no process matching '{pattern}'")
                return False
            for pid in pids:
                try:
                    os.kill(pid, signal.SIGTERM)
                    self.get_logger().info(f"[Watchdog] SIGTERM pid={pid} ({pattern})")
                except ProcessLookupError:
                    pass
            return True
        except Exception as e:
            self.get_logger().error(f"[Watchdog] kill failed: {e}")
            return False


def main():
    rclpy.init()
    node = TopicWatchdog()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()

File: physicar_bringup/src/topic_watchdog_node.cpp
/**
 * Topic Watchdog Node for PhysiCar (C++ port)
 *
 * Monitors critical sensor topics. If a topic stops publishing for longer
 * than its configured timeout, kills the responsible process (launch will
 * respawn it because every node has respawn=True).
 *
 * Restart policy:
 * - Startup grace period: ignore stale topics for the first N seconds.
 * - Cooldown: don't kill the same target twice within `cooldown` seconds.
 * - Sim mode: detect /clock backward jumps (world switch) and kill immediately.
 */

#include <chrono>
#include <csignal>
#include <cstdlib>
#include <fstream>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <rosgraph_msgs/msg/clock.hpp>
#include <sensor_msgs/msg/battery_state.hpp>
#include <sensor_msgs/msg/camera_info.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>

using namespace std::chrono;
using Clock = std::chrono::steady_clock;

struct WatchEntry
{
  std::string topic;
  double timeout;
  std::string kill_pattern;
};

// Real-mode watch list
static const std::vector<WatchEntry> REAL_WATCH_TOPICS = {
  {"/camera/camera_info", 5.0, "camera_ros/lib/camera_ros/camera_node"},
  {"/scan", 3.0, "rplidar_node"},
  {"/scan_filtered", 3.0, "scan_filter_node"},
  {"/odom", 3.0, "ekf_filter_node"},
  {"/battery_state", 15.0, "physicar_driver_node"},
  {"/imu", 5.0, "physicar_driver_node"},
};

// Sim-mode watch list
static const std::vector<WatchEntry> SIM_WATCH_TOPICS = {
  {"/scan", 10.0, "ros_gz_bridge"},
  {"/odom/laser", 5.0, "rf2o_laser_odometry_node"},
  {"/odom", 15.0, "ekf_filter_node"},
};

class TopicWatchdog : public rclcpp::Node
{
public:
  TopicWatchdog() : Node("topic_watchdog")
  {
    declare_parameter("startup_grace_sec", 30.0);
    declare_parameter("cooldown_sec", 30.0);
    declare_parameter("check_period_sec", 2.0);
    declare_parameter("enabled", true);
    declare_parameter("mode", std::string("real"));

    enabled_ = get_parameter("enabled").as_bool();
    if (!enabled_) {
      RCLCPP_INFO(get_logger(), "[Watchdog] disabled (enabled=False)");
      return;
    }

    auto mode = get_parameter("mode").as_string();
    const auto & entries = (mode == "sim") ? SIM_WATCH_TOPICS : REAL_WATCH_TOPICS;

    startup_grace_ = get_parameter("startup_grace_sec").as_double();
    cooldown_ = get_parameter("cooldown_sec").as_double();
    auto check_period = get_parameter("check_period_sec").as_double();

    start_time_ = Clock::now();
    my_pid_ = getpid();

    // Best-effort QoS for sensors
    auto qos_be = rclcpp::SensorDataQoS().keep_last(1);
    // Reliable QoS for other topics
    auto qos_rel = rclcpp::QoS(1).reliable();

    for (const auto & e : entries) {
      topics_.push_back(e);
      last_msg_time_[e.topic] = start_time_;

      // Choose QoS: best-effort for LaserScan and Imu topics
      bool is_sensor = (e.topic.find("/scan") != std::string::npos ||
                        e.topic == "/imu");
      auto & qos = is_sensor ? qos_be : qos_rel;

      // Generic subscription — we only need the callback, not the data
      subs_.push_back(create_generic_subscription(
        e.topic,
        get_msg_type(e.topic),
        qos,
        [this, topic = e.topic](std::shared_ptr<rclcpp::SerializedMessage>) {
          last_msg_time_[topic] = Clock::now();
        }));
    }

    timer_ = create_wall_timer(
      duration<double>(check_period),
      std::bind(&TopicWatchdog::check, this));

    RCLCPP_INFO(get_logger(),
      "[Watchdog] mode=%s, monitoring %zu topics (grace=%.0fs, cooldown=%.0fs)",
      mode.c_str(), topics_.size(), startup_grace_, cooldown_);

    // Sim mode: monitor /clock for backward time jumps
    if (mode == "sim") {
      clock_sub_ = create_subscription<rosgraph_msgs::msg::Clock>(
        "/clock", qos_be,
        std::bind(&TopicWatchdog::on_clock, this, std::placeholders::_1));
    }
  }

private:
  static std::string get_msg_type(const std::string & topic)
  {
    if (topic.find("/camera_info") != std::string::npos) return "sensor_msgs/msg/CameraInfo";
    if (topic.find("/scan") != std::string::npos) return "sensor_msgs/msg/LaserScan";
    if (topic == "/imu") return "sensor_msgs/msg/Imu";
    if (topic.find("/odom") != std::string::npos) return "nav_msgs/msg/Odometry";
    if (topic.find("/battery") != std::string::npos) return "sensor_msgs/msg/BatteryState";
    if (topic == "/clock") return "rosgraph_msgs/msg/Clock";
    return "std_msgs/msg/Empty";
  }

  void on_clock(const rosgraph_msgs::msg::Clock::SharedPtr msg)
  {
    double sec = msg->clock.sec + msg->clock.nanosec * 1e-9;
    double prev = last_sim_sec_;
    last_sim_sec_ = sec;

    if (prev < 1.0 || sec >= prev) return;

    auto now = Clock::now();
    if (duration<double>(now - start_time_).count() < startup_grace_) return;

    RCLCPP_WARN(get_logger(),
      "[Watchdog] /clock jumped backward (%.1fs -> %.1fs) — world switch detected",
      prev, sec);

    std::unordered_map<std::string, bool> killed;
    for (const auto & e : topics_) {
      if (killed.count(e.kill_pattern)) continue;
      auto it = last_kill_time_.find(e.kill_pattern);
      if (it != last_kill_time_.end() &&
          duration<double>(now - it->second).count() < cooldown_) continue;
      if (kill_process(e.kill_pattern)) {
        last_kill_time_[e.kill_pattern] = now;
        killed[e.kill_pattern] = true;
      }
    }
    for (const auto & e : topics_) {
      last_msg_time_[e.topic] = now;
    }
  }

  void check()
  {
    auto now = Clock::now();
    if (duration<double>(now - start_time_).count() < startup_grace_) return;

    for (const auto & e : topics_) {
      double stale = duration<double>(now - last_msg_time_[e.topic]).count();
      if (stale < e.timeout) continue;

      auto it = last_kill_time_.find(e.kill_pattern);
      if (it != last_kill_time_.end() &&
          duration<double>(now - it->second).count() < cooldown_) continue;

      RCLCPP_WARN(get_logger(),
        "[Watchdog] %s stale for %.1fs (timeout=%.0fs) -> killing '%s'",
        e.topic.c_str(), stale, e.timeout, e.kill_pattern.c_str());

      if (kill_process(e.kill_pattern)) {
        last_kill_time_[e.kill_pattern] = now;
        for (const auto & t : topics_) {
          if (t.kill_pattern == e.kill_pattern) {
            last_msg_time_[t.topic] = now;
          }
        }
      }
    }
  }

  static bool exe_matches(pid_t pid, const std::string & pattern)
  {
    // Check /proc/PID/exe (binary path)
    char buf[512];
    std::string link = "/proc/" + std::to_string(pid) + "/exe";
    ssize_t len = ::readlink(link.c_str(), buf, sizeof(buf) - 1);
    if (len > 0) {
      buf[len] = '\0';
      if (std::string(buf).find(pattern) != std::string::npos) return true;
    }
    // Fallback: check /proc/PID/cmdline (includes node name remapping etc.)
    std::string cmdline_path = "/proc/" + std::to_string(pid) + "/cmdline";
    std::ifstream ifs(cmdline_path);
    if (ifs) {
      std::string cmdline((std::istreambuf_iterator<char>(ifs)),
                           std::istreambuf_iterator<char>());
      if (cmdline.find(pattern) != std::string::npos) return true;
    }
    return false;
  }

  bool kill_process(const std::string & pattern)
  {
    // Use pgrep to find candidate PIDs, then verify via /proc/PID/exe
    std::string cmd = "pgrep -f '" + pattern + "' 2>/dev/null";
    FILE * fp = popen(cmd.c_str(), "r");
    if (!fp) return false;

    std::vector<pid_t> pids;
    char buf[32];
    while (fgets(buf, sizeof(buf), fp)) {
      pid_t pid = static_cast<pid_t>(std::atoi(buf));
      if (pid > 0 && pid != my_pid_ && exe_matches(pid, pattern)) {
        pids.push_back(pid);
      }
    }
    pclose(fp);

    if (pids.empty()) {
      RCLCPP_WARN(get_logger(), "[Watchdog] no process matching '%s'", pattern.c_str());
      return false;
    }

    for (pid_t pid : pids) {
      if (::kill(pid, SIGTERM) == 0) {
        RCLCPP_INFO(get_logger(), "[Watchdog] SIGTERM pid=%d (%s)", pid, pattern.c_str());
      }
    }
    return true;
  }

  bool enabled_;
  double startup_grace_;
  double cooldown_;
  pid_t my_pid_;
  Clock::time_point start_time_;
  double last_sim_sec_ = 0.0;

  std::vector<WatchEntry> topics_;
  std::unordered_map<std::string, Clock::time_point> last_msg_time_;
  std::unordered_map<std::string, Clock::time_point> last_kill_time_;

  std::vector<std::shared_ptr<rclcpp::GenericSubscription>> subs_;
  rclcpp::Subscription<rosgraph_msgs::msg::Clock>::SharedPtr clock_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TopicWatchdog>());
  rclcpp::shutdown();
  return 0;
}
